### ISSUE TITLE
Server Side Request Forgery via Import Site feature (related to closed issue #1026)

### ISSUE THREAD

**Issue by dunderhay** [gophish/gophish#1908]
### What version of Gophish are you using?:
Gophish v0.10.1

### Brief description of the issue:
This issue has previously been raised [here #1026](https://github.com/gophish/gophish/issues/1026). It appears there were actually two seperate issues reported originally:

1) Lack of Authentication on endpoints (/import/site , /import/email and /import/group).
2) Server Side Request Forgery (SSRF)

The fix for the authentication was implemented but the SSRF is still exploitable and appears that there is no fix for this (it is functionality acting as intended).  By nature, the import site functionality needs to be able to make requests to any external IP address or domain name in order to clone the site. So an allow listing approach is not really feasible. Perhaps a regex to block requests to localhost, internal IP address ranges and AWS endpoints, but this just seems like a never ending battle? 🤷‍♂️

The account for the person who originally reported the issue appears to have been deleted but as per their issue:
> It is possible to access the "Import Site" endpoint of the "Create Landing Page" ~~area without cookie/API key, AKA without authentication~~. ~~As such unauthenticated~~ users who can access the admin portal can use it to perform server side request forgery attacks. Note: this is only really an issue if the admin interface is listening on anything other than 127.0.0.1, or if it is running on 127.0.0.1 on a multiuser machine accessed by non gophish admins.

I have struck out the text which is no longer relevant to reflect the authentication updates which have been made. The endpoints now require authentication but it is still possible to abuse the SSRF vulnerability to achieve the following:
- Attacks against the server itself (target other services running on localhost)
- Internal port scanning
- Attacks against other back-end systems e.g:
   - Abusing trust relationship between gophish and mail server
   - Querying AWS metadata service

__Note:__ AWS has added support for IMDSv2 to mitigate SSRF attacks against AWS metadata service. Also I have checked if various protocols and URL schemas (ssh, smb, file, gopher, dict, etc) and had no luck with any of these being supported.

### What are you expecting to see happen?
Expecting to not be able to port scan localhost, internal IP address ranges, and access AWS metadata service.

### What are you seeing happen?
It is possible to query the AWS metadata service and port scan localhost / internal IP address ranges. For example, it is possible to fetch the EC2 instance identity document (`http://169.254.169.254/latest/dynamic/instance-identity/document`) by inserting this as the endpoint to attempt to clone in the "Import Site" feature.
 
![ssrf-req](https://user-images.githubusercontent.com/27876907/88356382-f82c2b00-cdaa-11ea-926a-56561e86ac10.png)
![ssrf-resp](https://user-images.githubusercontent.com/27876907/88356384-fa8e8500-cdaa-11ea-8062-eaf8500ed659.png)

**Comment by jordan-wright** [comment]
Hi there!

I'll get back to this as soon as I can, but in the future I'd ask that you please send any future security issues to hi@getgophish.com rather than creating a public issue.

Thanks for getting in touch!

**Comment by jordan-wright** [comment]
Ok, thanks for the patience.

I'll say that this was also reported a couple of weeks ago via email and since then I've been thinking about how to handle this. Normally, SSRF issues are straightforward since you don't expect internal URLs or services to be hit. Just restrict those and problem (largely) solved. But in Gophish, that's simply not the case.

There have been multiple folks in the issues that are trying to import landing pages from localhost. And while that is _ideally_ not the most common thing, it would definitely be expected that people may want to import landing pages that are on the internal network.

Any place where we allow network connections to be established from Gophish, we expect internal services to be candidates for that. That's makes it difficult to implement good SSRF protection.

I think there's an argument that importing sites is the most dangerous here since we actual render the returned HTML, vs. other places where maybe we only try to establish a connection and give a 👍 or 👎 . 

So at this point, we pretty much know that a blocklist is probably the route we need to go. But we also know that internal services, localhost, etc. are all valid to be requested. So we could go the route of blocking very, very few things (like metadata URLs) but it just feels like it's extra effort and it'll always be a cat/mouse kind of scenario.

Alternatively, we could take the route of blocking access to internal services by default, and allowing an override via an environment variable or something. It's more effort, but it'd be possible.

Curious to get your thoughts on this.

**Comment by dunderhay** [comment]
Hey Jordan,

Yeah agree with you that in this use case, the SSRF is really tricky to try prevent. Also valid input that folks might want to clone their landing page from an internal resource (never done this myself personally).

Like you said, the blocklist approach is very cat/mouse and ends up being a lot of work. I do like the second suggestion you mentioned (environment variable or setting to allow fetching resources from *dangerous* endpoints - or something like that?).

**Comment by jordan-wright** [comment]
Hey @dunderhay,

I've opted for a slightly different approach. Since it's very expected behavior for Gophish to contact internal hosts, I've opted to have a very small denylist by default (e.g. only the common link-local IP address `169.254.169.254` used for instance metadata), and then made a configuration option that can be set to restrict access to certain address ranges, which I'll ensure is listed in the documentation.

I'll be releasing a new version of Gophish soon that includes this fix.
